### PR TITLE
Unlocking Fixes

### LINKED ISSUE #922: Ordered unlocking within groups
Ability to set a puzzle to unlock once X other puzzles in its group have unlocked. Note: this may require making a table for groups (they're currently just a string in the Puzzles table).

### LINKED ISSUE #938: Change unlock time for zero-prerequisite puzzles to EventBegin time
PuzzleStateHelper.CheckForTimedUnlocksAsync, change

                `await PuzzleStateHelper.SetUnlockStateAsync(context, eventObj, puzzle, team, now);`

to

                `await PuzzleStateHelper.SetUnlockStateAsync(context, eventObj, puzzle, team, eventObj.EventBegin);`

### PR DESCRIPTION
**Pull request by tabascq**
Adds a MinInGroupCount that can be used to unlock puzzles when enough puzzles in the Group have been solved.

Note that MinInGroupCount is nullable and the original MinPrerequisiteCount is not, so if you have a puzzle that you want to unlock based on group solves, you should set MinPrerequisiteCount to be some positive integer so it doesn't automatically unlock based on prerequisites. You can also set timed puzzles that count as prerequisites, and you should no longer need to make multiple copies of the same timer.

Also fixes the solve time of zero-prerequisite puzzles to be the event begin time instead of the current time.

Fixes #922.
Fixes #938.

### CODE REVIEW

File: ServerCore/PuzzleStateHelper.cs [hunk 1]
```diff
@@ -336,7 +336,7 @@ public static async Task<bool> IsPuzzleSolved(PuzzleServerContext context, int p
                 .ToListAsync();
             foreach (var puzzle in zeroPrerequisitePuzzlesToUnlock)
             {
-                await PuzzleStateHelper.SetUnlockStateAsync(context, eventObj, puzzle, team, now);
+                await PuzzleStateHelper.SetUnlockStateAsync(context, eventObj, puzzle, team, eventObj.EventBegin);
```
[the remark this answers is not in the document]
Comment (digit01Wave) [reply ServerCore/PuzzleStateHelper.cs:338]: Why did this update to evetnObj.EventBegin?

[the remark this answers is not in the document]
Comment (tabascq) [reply ServerCore/PuzzleStateHelper.cs:338]: There was a request to do this, which is covered in #938.

File: ServerCore/Pages/Puzzles/Edit.cshtml [hunk 2]
```diff
@@ -98,6 +98,11 @@
                                 <input asp-for="Puzzle.MinPrerequisiteCount" class="form-control" />
                                 <span asp-validation-for="Puzzle.MinPrerequisiteCount" class="text-danger"></span>
                             </div>
+                            <div class="form-group">
```
Comment (morganbr) [inline ServerCore/Pages/Puzzles/Edit.cshtml:101]: We probably also need this on the create page

Comment (morganbr) [inline ServerCore/Pages/Puzzles/Edit.cshtml:101]: Does this input setup allow you to go to/from null?

**Review by morganbr — CHANGES_REQUESTED** [review]

**Review by morganbr — APPROVED** [review]

**Pull request merged**